File: job-recommender-frontend/src/app/services/job-posting.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class JobPostingService {
  private baseUrl = 'http://localhost:8080/api/job-postings'; // Base URL for job postings API

  constructor(private http: HttpClient) {}


  getPagedJobPostings(page: number, size: number): Observable<any> {
    return this.http.get<any>(
      `${this.baseUrl}/paged?page=${page}&size=${size}`
    );
  }
  // Fetch all job postings
  getAllJobPostings(): Observable<any> {
    return this.http.get(`${this.baseUrl}`);
  }

  // Fetch a single job posting by ID
  getJobPostingById(id: number): Observable<any> {
    return this.http.get(`${this.baseUrl}/${id}`);
  }

  // Create a new job posting
  createJobPosting(jobPosting: any): Observable<any> {
    return this.http.post(`${this.baseUrl}`, jobPosting);
  }

  // Update an existing job posting
  updateJobPosting(id: number, jobPosting: any): Observable<any> {
    return this.http.put(`${this.baseUrl}/${id}`, jobPosting);
  }

  // Delete a job posting by ID
  deleteJobPosting(id: number): Observable<any> {
    return this.http.delete(`${this.baseUrl}/${id}`);
  }
}

File: TestPFA/src/main/java/com/example/TestPFA/service/JobPostingService.java
package com.example.TestPFA.service;

import com.example.TestPFA.entity.JobPosting;
import com.example.TestPFA.repository.JobPostingRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class JobPostingService {

    private final JobPostingRepository jobPostingRepository;

    // Constructor injection of the repository
    public JobPostingService(JobPostingRepository jobPostingRepository) {
        this.jobPostingRepository = jobPostingRepository;
    }

    // Example: Retrieve all job postings
    public List<JobPosting> getAllJobPostings() {
        return jobPostingRepository.findAll();
    }

    // Example: Save a new job posting
    public JobPosting saveJobPosting(JobPosting jobPosting) {
        return jobPostingRepository.save(jobPosting);
    }

    // Example: Retrieve a single job posting by ID
    public JobPosting getJobPostingById(Long id) {
        return jobPostingRepository.findById(id)
                .orElse(null); // or throw an exception if not found
    }

    // Example: Delete a job posting by ID
    public void deleteJobPosting(Long id) {
        jobPostingRepository.deleteById(id);
    }

    // Update an existing job posting
    public JobPosting updateJobPosting(Long id, JobPosting updatedJobPosting) {
        return jobPostingRepository.findById(id)
                .map(existing -> {
                    existing.setJobTitle(updatedJobPosting.getJobTitle());
                    existing.setJobDescription(updatedJobPosting.getJobDescription());
                    existing.setJobType(updatedJobPosting.getJobType());
                    existing.setLocation(updatedJobPosting.getLocation());
                    existing.setOrganization(updatedJobPosting.getOrganization());
                    existing.setSector(updatedJobPosting.getSector());
                    existing.setCity(updatedJobPosting.getCity());
                    existing.setStateCountry(updatedJobPosting.getStateCountry());
                    return jobPostingRepository.save(existing);
                })
                .orElse(null);
    }
    public List<JobPosting> getPostingsByJobType(String jobType) {
        // Assuming you create a custom query method in the repository:
        // List<JobPosting> findByJobType(String jobType);

        return jobPostingRepository.findByJobType(jobType);
    }
    /*
     * In the future, we’ll add methods for "recommendations"
     * (e.g., filtering logic based on job type, skills, etc.)
     * For now, let's keep the basic CRUD.
     */
}